File: backend/src/models/ticket.py
from beanie import Document,Link, PydanticObjectId
from pydantic import Field
from typing import Optional, List, Dict
from datetime import datetime
from .category import Category
from .subcategory import SubCategory
from .enums import TicketStatus, TicketPriority, TicketSeverity
from .user import User
class Ticket(Document):
    id: Optional[PydanticObjectId] = Field(default=None, alias="_id", description="Primary key (MongoDB ObjectId)")
    categoryId: Link[Category] = Field(..., description="ID of the category this ticket belongs to")
    subCategoryId: Link[SubCategory] = Field(..., description="ID of the subcategory this ticket belongs to")
    userId: Link[User] = Field(..., description="ID of the user who created the ticket")
    agentId: Optional[Link[User]] = Field(None, description="ID of the agent assigned to the ticket (if any)")
    
    title: str = Field(..., description="Title of the ticket")
    description: str = Field(..., description="Detailed description of the ticket")
    content: str=  Field(..., description="Content of the ticket, can be rich text")
    tagIds: Optional[List[Dict[str, str]]] = Field(default_factory=list, description="List of tag IDs associated with the ticket")

    status: TicketStatus = Field(default=TicketStatus.open, description="Status of the ticket") 
    priority: TicketPriority = Field(default=TicketPriority.medium, description="Priority: low, medium, high")
    severity: TicketSeverity = Field(default=TicketSeverity.low, description="Severity of the issue")

    createdAt: datetime = Field(default_factory=datetime.utcnow)
    updatedAt: datetime = Field(default_factory=datetime.utcnow)
    closedAt: Optional[datetime] = None

    class Settings:
        name = "tickets"  # MongoDB collection name

    

File: frontend/src/app/shared/types/ticket.ts
// Ticket-related types matching backend API
import { Category } from './category';
import { SubCategory } from './subcategory';

export type TicketStatus = 'new' | 'in_progress' | 'waiting_for_customer' | 'waiting_for_agent' | 'resolved' | 'closed';
export type TicketPriority = 'low' | 'medium' | 'high' | 'critical';

// Rich text content structure
export interface RichTextContent {
  html: string;
  json: Record<string, unknown>;
  text: string;
}

export interface Ticket {
  id: string;
  title: string;
  description: string;
  content: RichTextContent;
  category: Category;
  subCategory: SubCategory;
  userInfo: UserInfo;
  agentInfo?: UserInfo;
  tagIds?: Array<{ key: string; value: string }>;
  status: TicketStatus;
  priority: TicketPriority;
  createdAt: string;
  updatedAt: string;
  closedAt?: string;
  // ENHANCEMENT L1 AI TICKET SUMMARY - Add AI summary fields
  aiSummary?: string;
  summaryGeneratedAt?: string;
}

export interface UserInfo {
  id: string;
  email: string;
  name?: string;
  role?: string; // "user" or "agent" - helps identify user type in comments and tickets
}

export interface CreateTicketRequest {
  category_id: string;
  sub_category_id: string;
  title: string;
  description: string;
  content: RichTextContent;
  priority?: TicketPriority;
  tag_ids?: { [key: string]: string }[];
}

export interface UpdateTicketRequest {
  title?: string;
  description?: string;
  content?: RichTextContent;
  status?: TicketStatus;
  priority?: TicketPriority;
  agent_id?: string;
  tag_ids?: { [key: string]: string }[];
}

export interface TicketStats {
  total: number;
  new: number;
  in_progress: number;
  waiting_for_customer: number;
  waiting_for_agent: number;
  resolved: number;
  closed: number;
  by_priority: {
    low: number;
    medium: number;
    high: number;
    critical: number;
  };
}

export interface Comment {
  id: string;
  ticket_id: string;
  user: UserInfo;
  content: RichTextContent;
  created_at: string;
  updated_at: string;
}

export interface CreateComment {
  content: RichTextContent;
}

export interface UpdateComment {
  content: RichTextContent;
}

// ENHANCEMENT L1 AI TICKET SUMMARY - AI summary types
export interface TicketSummaryResponse {
  summary: string;
}

export interface ClosingCommentsResponse {
  suggestions: string[];
}

// Legacy aliases for backward compatibility
export type CreateTicket = CreateTicketRequest;
export type UpdateTicket = UpdateTicketRequest;

export interface TicketAssignment {
  agent_id: string;
}

export interface TicketClosure {
  resolution: string;
}